kernel/vmbytegen.ml

```diff
@@ -799,37 +815,39 @@ let rec compile_lam env cenv lam sz cont =
 
 and compile_get_global env cenv (kn,u) sz cont =
   let () = set_max_stack_size cenv sz in
-  if Univ.Instance.is_empty u then
+  if UVars.Instance.is_empty u then
     Kgetglobal kn :: cont
   else
     comp_app (fun _ _ _ cont -> Kgetglobal kn :: cont)
       (compile_instance env) cenv () [|u|] sz cont
 
-and compile_instance env cenv u0 sz cont =
+and compile_instance env cenv u sz cont =
   let () = set_max_stack_size cenv sz in
-  let u = Univ.Instance.to_array u0 in
-  let len = Array.length u in
-  let is_id i l = match Univ.Level.var_index l with
-  | None -> false
-  | Some j -> Int.equal i j
-  in
-  if Int.equal env.uinst_len len && Array.for_all_i is_id 0 u then
+  if is_toplevel_inst env u then
     (* Optimization: do not reallocate the same instance *)
     pos_instance cenv sz :: cont
-  else if Array.for_all (fun l -> Option.is_empty (Univ.Level.var_index l)) u then
+  else if is_closed_inst u then
     (* Optimization: allocate closed instances globally *)
-    compile_structured_constant cenv (Const_univ_instance u0) sz cont
+    compile_structured_constant cenv (Const_univ_instance u) sz cont
   else
-    let () = set_max_stack_size cenv (sz + len - 1) in
+    let qs, us = UVars.Instance.to_array u in
+    let qlen, ulen = UVars.Instance.length u in
+    let () = set_max_stack_size cenv (sz + qlen + ulen - 1) in
+    let comp_qual cenv q sz cont = match Sorts.Quality.var_index q with
+    | None -> compile_structured_constant cenv (Const_quality q) sz cont
+    | Some idx -> pos_instance cenv sz :: Kfield 0 :: Kfield idx :: cont
+    in
     let comp_univ cenv l sz cont = match Univ.Level.var_index l with
     | None -> compile_structured_constant cenv (Const_univ_level l) sz cont
-    | Some idx -> pos_instance cenv sz :: Kfield idx :: cont
+    | Some idx -> pos_instance cenv sz :: Kfield 1 :: Kfield idx :: cont
     in
-    comp_args comp_univ cenv u sz (Kmakeblock (len, 0) :: cont)
+    let cont = Kmakeblock (2, 0) :: cont in
+    let cont = comp_args comp_univ cenv us sz (Kmakeblock (ulen, 0) :: cont) in
+    comp_args comp_qual cenv qs sz (Kmakeblock (qlen, 0) :: cont)
```

Comment:
Something around here seems bugged but I don't understand what I'm doing enough to tell how to fix it. It segfaults on 
~~~
Set Universe Polymorphism.

Definition bar@{i} := Type@{i}.

Definition foo@{i j} := bar@{i}.

Eval vm_compute in foo.
~~~
(also the examples from #7723) 

The change is that an instance was an array of univ levels and is now a pair of an array of sort qualities and an array of univ levels.

cc @ppedrot @silene

Reply:
Maybe the order of comp_univ / comp_qual is wrong but switching it doesn't fix the segfault so IDK

Reply:
Your stack tracking is presumably wrong. I would expect `(sz + 1)` on the second call to `comp_args`.

Also, your changes to `comp_args` seem a bit dangerous. Since the function is meant to push `n - 1` arguments on the stack and one in the accumulator, if you have `n = 0`, what do you write in the accumulator?

Reply:
And there should be a `Kpush` inbetween. (Otherwise the stack tracking is actually correct, but the content of the stack then does not make sense to me.)

Reply:
If you mean
~~~ocaml
    let cont = comp_args comp_qual cenv qs sz (Kmakeblock (qlen,0) :: cont) in
    comp_args comp_univ cenv us sz (Kmakeblock (ulen,0) :: Kpush :: cont)
~~~
that still segfaults

Reply:
As I said, you need to track the stack size, so
```ocaml
    let cont = comp_args comp_qual cenv qs (sz + 1) (Kmakeblock (qlen,0) :: cont) in
    comp_args comp_univ cenv us sz (Kmakeblock (ulen,0) :: Kpush :: cont)
```

Reply:
Thanks, it seems to work now.

>Also, your changes to comp_args seem a bit dangerous.

I moved them to a local wrapper.